src/main/java/com/linecorp/armeria/server/http/tomcat/TESTInvocationHandler.java

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright 2015 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.http.tomcat;
+
+import java.net.InetSocketAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.Map.Entry;
+import java.util.concurrent.Executor;
+
+import org.apache.catalina.connector.Connector;
+import org.apache.coyote.Adapter;
+import org.apache.coyote.InputBuffer;
+import org.apache.coyote.OutputBuffer;
+import org.apache.coyote.Request;
+import org.apache.coyote.Response;
+import org.apache.tomcat.util.buf.ByteChunk;
+import org.apache.tomcat.util.buf.MessageBytes;
+import org.apache.tomcat.util.http.MimeHeaders;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.ServiceInvocationContext;
+import com.linecorp.armeria.server.ServiceInvocationHandler;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.handler.codec.http.DefaultFullHttpResponse;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.util.AsciiString;
+import io.netty.util.concurrent.Promise;
+
+final class TESTInvocationHandler implements ServiceInvocationHandler {
```

Comment:
> about extending : TESTInvocationHandler(new) assumes tomcat is already configured and running. but TomcatServiceInvocationHandler(existing) does not guarantee this. this makes hard to extend.

I'm not sure I understood your concern. The existing TomcatServiceInvocationHandler could extend TESTInvocationHandler and implement handlerAdded()? 

> also my approach lacks some of life-cycle management, especially closing. it does not cleanup armeria server even though tomcat is being shutdown. do you have any suggestion on this?

That's a good point. Perhaps we should start responding with a 503 Service Unavailable reponse when Tomcat starts to shut down?